Re-implement spectral-cube methods with dask

astrofrog opened the pull request:
I started working on https://github.com/radio-astro-tools/spectral-cube/pull/567 but I found that we can probably implement dask support in a simpler way that is not mixed in with the joblib-related code, so I decided to start a new PR and go through and systematically implement all methods/properties.

Here are all the methods/properties that exist on ``SpectralCube``, and I'll check them if they are done or score them out if they don't need to be overridden:

<details>

* [x] apply_function
* [x] apply_function_parallel_spatial
* [x] apply_function_parallel_spectral
* [x] apply_numpy_function
* [x] argmax
* [x] argmin
* [x] <s>base</s> (not needed, only existed for joblib)
* [x] <s>beam</s> (unrelated to data array)
* [x] <s>chunked</s> (``NotImplemented`` for SpectralCube)
* [x] <s>closest_spectral_channel</s> (unrelated to data array)
* [x] convolve_to
* [x] downsample_axis
* [x] <s>fill_value</s> (unrelated to data array)
* [ ] filled
* [x] filled_data
* [x] <s>find_lines</s> (unrelated to data array)
* [x] flattened
* [x] <s>flattened_world</s> (wrapper around ``world``)
* [x] get_mask_array
* [x] hdu
* [x] <s>hdulist</s> (thin wrapper around ``hdu``)
* [x] <s>header</s>
* [x] <s>latitude_extrema</s> (wrapper around ``world_extrema``)
* [x] <s>linewidth_fwhm</s> (wrapper around ``linewidth_sigma``)
* [x] <s>linewidth_sigma</s> (wrapper around ``moment2``)
* [x] <s>longitude_extrema</s> (wrapper around ``world_extrema``)
* [x] mad_std
* [x] <s>mask</s> (unrelated to data array)
* [x] <s>mask_channels</s> (wrapper around ``with_mask``)
* [x] max
* [x] mean
* [x] median
* [x] <s>meta</s> (unrelated to data array)
* [x] min
* [x] <s>minimal_subcube</s> (wrapper around ``subcube_slices_from_mask`` and ``__getitem__``)
* [x] moment
* [x] <s>moment0</s> (wrapper around ``moment``)
* [x] <s>moment1</s> (wrapper around ``moment``)
* [x] <s>moment2</s> (wrapper around ``moment``)
* [x] <s>ndim</s> (works because dask array exposes this too)
* [x] percentile
* [x] <s>pixels_per_beam</s> (unrelated to data array)
* [x] plot_channel_maps
* [x] read
* [ ] reproject
* [x] <s>shape</s> (works because dask array exposes this too)
* [x] sigma_clip_spectrally
* [x] <s>size</s> (works because dask array exposes this too)
* [x] <s>spatial_coordinate_map</s> (wrapper around ``world``)
* [x] spatial_smooth
* [x] spatial_smooth_median
* [x] <s>spectral_axis</s> (wrapper around ``world``)
* [x] <s>spectral_extrema</s> (wrapper around ``spectral_axis``)
* [x] spectral_interpolate
* [x] <s>spectral_slab</s> (only uses slicing which should work with dask)
* [x] spectral_smooth
* [x] spectral_smooth_median
* [x] std
* [x] <s>subcube</s> (wrapper around ``__getitem__``)
* [x] subcube_from_crtfregion
* [x] subcube_from_ds9region
* [x] <s>subcube_from_mask</s> (wrapper around ``subcube_slices_from_mask``)
* [x] subcube_from_regions
* [x] subcube_slices_from_mask
* [x] sum
* [x] <s>to</s> (wrapper around ``_new_cube_with``)
* [ ] to_ds9
* [x] <s>to_glue</s> (works as-is)
* [x] <s>to_pvextractor</s> (works as-is)
* [ ] to_yt
* [x] <s>unit</s> (unrelated to data array)
* [x] <s>unitless</s> (wrapper around ``_new_cube_with``)
* [x] unitless_filled_data
* [x] unmasked_copy
* [x] unmasked_data
* [x] <s>velocity_convention</s> (unrelated to data array)
* [x] <s>wcs</s> (unrelated to data array)
* [x] <s>with_beam</s> (keeps data as-is)
* [x] <s>with_fill_value</s> (keeps data as-is)
* [x] <s>with_mask</s> (keeps data as-is)
* [x] <s>with_spectral_unit</s> (keeps data as-is)
* [x] world
* [x] <s>world_extrema</s> (wrapper around ``shape`` and ``wcs``)
* [x] world_spines (``NotImplemented`` for SpectralCube)
* [x] write

</details>

In addition to this, I will need to:

* [x] Make sure ``DaskSpectralCube.read/write`` work as expected
* [x] Determine how to deal with the varying beam spectral cube class
* [x] Update tests to try both the plain and the dask class
* [x] Do systematic benchmarking of performance
* [x] Make sure parallel versions are efficient

keflavich commented:
I'm trying to test this, and no matter what kind of cube (`SpectralCube` or `DaskSpectralCube`) I tried, I can't get simple things to work for this large cube:
```
In [2]: from spectral_cube.dask_spectral_cube import DaskSpectralCube                                                                                          

In [3]: cube = DaskSpectralCube.read('BrickMaser_95_spw25.residual',       format='casa_image')                                                                

In [4]: cube                                                                                                                                                   
Out[4]: 
SpectralCube with shape=(16127, 540, 540):
 n_x:    540  type_x: RA---SIN  unit_x: deg    range:   266.522982 deg:  266.565664 deg
 n_y:    540  type_y: DEC--SIN  unit_y: deg    range:   -28.723679 deg:  -28.686248 deg
 n_s:  16127  type_s: FREQ      unit_s: Hz     range: 93550495256.267 Hz:109296464727.664 Hz

In [5]: maxspec = cube.max(axis=(1,2))                                                                                                                         
Killed
```

I had expected dask to help with this - any idea why it's not helping?  (this may be more appropriate as its own issue, but... still... dask was supposed to save us!)

keflavich commented:
Possible UI breakage introduced here (?):
```
In [3]: cube[8000,:,:].quicklook()                                                                                                                             
Traceback (most recent call last):
  File "<ipython-input-3-5ba6f8023869>", line 1, in <module>
    cube[8000,:,:].quicklook()
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/spectral_cube.py", line 1242, in __getitem__
    mask=self.mask[view],
TypeError: 'NoneType' object is not subscriptable
```
(again, this might be a CASA reader issue rather than a dask issue)

e-koch commented:
@keflavich - Does the residual cube have a mask? The reader may need a default mask when none is found.

keflavich commented:
@e-koch let's move this discussion to #619

astrofrog commented:
@keflavich - could you try the ``max()`` example again? The issue was that ``DaskSpectralCube.read`` was returning a ``SpectralCube`` (which is not optimized for dask) rather than a ``DaskSpectralCube``. With the example 68,000 channel cube you sent me, max takes <2 minutes.

keflavich commented:
```
  File "<ipython-input-4-62fca8ad7498>", line 12, in <module>
    cube = DaskSpectralCube.read(fn)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py", line 481, in read
    raise TypeError('SpectralCube._data is not a dask array')
TypeError: SpectralCube._data is not a dask array
```

keflavich commented:
My [error](https://github.com/radio-astro-tools/spectral-cube/pull/618#issuecomment-609950848) is from using a FITS-read cube instead of a CASA-read cube.

keflavich commented:
The return types are different from `SpectralCube` - no `OneDSpec` or `Projection` is returned, just an array.

astrofrog commented:
@keflavich - yep I haven't incorporated the Dask FITS reading here yet, and we're still far from feature parity - I was just curious whether the ``max()`` example you showed before worked since that should work now.

keflavich commented:
it worked; gave me an array.  👍

astrofrog commented:
@keflavich - I've now fixed the return types, so that ``DaskSpectralCube`` should now return the same as ``SpectralCube`` (and same for the variable resolution classes). I've also fixed the FITS I/O so that you can read in FITS files with ``DaskSpectralCube``. Anything ticked in the list above should work properly I think, and unticked things almost certainly **won't** work properly. Could you take this PR for a spin?

astrofrog commented:
@e-koch - if you fancy taking this for a spin, any feedback so far would be useful! (starting from ``from spectral_cube.dask_spectral_cube import DaskSpectralCube `` and ``DaskSpectralCube.read(...)``. As noted in my previous comment, unchecked items in the list above won't work well/efficiently.

e-koch commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,794 @@
+
+    @warn_slow
+    @projection_if_needed
+    def mad_std(self, axis=None, **kwargs):
```
`kwargs` aren't passed to `mad_std` below.

e-koch replied:
Should `ignore_nan=True` by default? I guess we didn't have this in `SpectralCube` but seems like a reasonable default since the rest of the operations ignore nans, too.

I'll raise a separate issue.

astrofrog replied:
Fixed!

e-koch commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,794 @@
+            raise ImportError("Scipy could not be imported: this function won't work.")
+
+        def median_filter_wrapper(img, **kwargs):
+            return ndimage.median_filter(img, (ksize, 1, 1), **kwargs)
```
If `ksize` is not an `int`, this fails with a `TypeError`. A type conversion or check with an helpful error message would be useful.  Somewhere inputting a `float` in the `SpectralCube` version is corrected for, though I don't see where...

astrofrog replied:
I think it's because here we pass a tuple of values not a scalar, which maybe causes issues.

e-koch commented:
@astrofrog - Reading in/etc is working great.  I worked through the list of implemented function and added comments for a couple small issues.

Otherwise, I had problems when `argmax` and `argmin` encountered an all NaN slice. It looks like this is a difference in how the dask functions handle NaNs along axes.
```
/home/ekoch/anaconda3/lib/python3.6/site-packages/dask/array/reductions.py:881: RuntimeWarning: All-NaN slice encountered
  vals = func(x, axis=arg_axis, keepdims=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-36-472eb331f44d> in <module>
----> 1 peakvel = cube.argmax(axis=0)

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/utils.py in wrapper(self, *args, **kwargs)
     47                           PossiblySlowWarning
     48                          )
---> 49         return function(self, *args, **kwargs)
     50     return wrapper
     51 

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/dask_spectral_cube.py in argmax(self, axis, **kwargs)
    221         excluded from the mask.
    222         """
--> 223         return self._compute(da.nanargmax(self._nan_filled_dask_array, axis=axis))
    224 
    225     @warn_slow

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/dask_spectral_cube.py in _compute(self, array)
    107         # For now, always default to serial mode, but we could then expand this
    108         # to allow different modes.
--> 109         return array.compute(scheduler='synchronous')
    110 
    111     @property

~/anaconda3/lib/python3.6/site-packages/dask/base.py in compute(self, **kwargs)
    164         dask.base.compute
    165         """
--> 166         (result,) = compute(self, traverse=False, **kwargs)
    167         return result
    168 

~/anaconda3/lib/python3.6/site-packages/dask/base.py in compute(*args, **kwargs)
    435     keys = [x.__dask_keys__() for x in collections]
    436     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 437     results = schedule(dsk, keys, **kwargs)
    438     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    439 

~/anaconda3/lib/python3.6/site-packages/dask/local.py in get_sync(dsk, keys, **kwargs)
    525     """
    526     kwargs.pop("num_workers", None)  # if num_workers present, remove it
--> 527     return get_async(apply_sync, 1, dsk, keys, **kwargs)
    528 
    529 

~/anaconda3/lib/python3.6/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    469             # Seed initial tasks into the thread pool
    470             while state["ready"] and len(state["running"]) < num_workers:
--> 471                 fire_task()
    472 
    473             # Main loop, wait on tasks to finish, insert new ones

~/anaconda3/lib/python3.6/site-packages/dask/local.py in fire_task()
    464                         pack_exception,
    465                     ),
--> 466                     callback=queue.put,
    467                 )
    468 

~/anaconda3/lib/python3.6/site-packages/dask/local.py in apply_sync(func, args, kwds, callback)
    514 def apply_sync(func, args=(), kwds={}, callback=None):
    515     """ A naive synchronous version of apply_async """
--> 516     res = func(*args, **kwds)
    517     if callback is not None:
    518         callback(res)

~/anaconda3/lib/python3.6/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    225         failed = False
    226     except BaseException as e:
--> 227         result = pack_exception(e, dumps)
    228         failed = True
    229     return key, result, failed

~/anaconda3/lib/python3.6/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    220     try:
    221         task, data = loads(task_info)
--> 222         result = _execute_task(task, data)
    223         id = get_id()
    224         result = dumps((result, id))

~/anaconda3/lib/python3.6/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

~/anaconda3/lib/python3.6/site-packages/toolz/functoolz.py in __call__(self, *args, **kwargs)
    466         ret = self.first(*args, **kwargs)
    467         for f in self.funcs:
--> 468             ret = f(ret)
    469         return ret
    470 

~/anaconda3/lib/python3.6/site-packages/dask/array/reductions.py in nanarg_agg(func, argfunc, data, axis, **kwargs)
    921     arg, vals = _arg_combine(data, axis, argfunc, keepdims=False)
    922     if np.any(np.isnan(vals)):
--> 923         raise ValueError("All NaN slice encountered")
    924     return arg
    925 

ValueError: All NaN slice encountered
```
It would be nice if we could make the `arg` functions default to 0 like the numpy equivalents, even if that's not actually "correct."

astrofrog commented:
@e-koch - I've fixed the argmin/argmax issue. Thanks for testing all this!

astrofrog commented:
I think this PR is now in good shape. I still need to go through and reduce any code duplication between the dask and base classes, but I think otherwise things should work properly. I've added docs about using multi-threading and controlling the number of workers. I get really good multi-threaded performance with FITS cubes, but not so much with CASA datasets, due I think to the very high number of chunks.

So things that remain to be done for this PR:

* [ ] More user testing (@keflavich @e-koch!)
* [ ] Clean up code to reduce duplication
* [ ] Add changelog entry

I've done testing of the performance on this PR and managed to get most methods to run for larger-than-memory cubes.

An additional item I want to work on, but in a separate PR, is to make it so that the cubes can be temporarily dumped to e.g. a zarr dataset to trigger computation and avoid long chains of computations that might go spectral -> image -> spectral and would be quite inefficient for dask to do all at once.

For now, trying out this PR would be most helpful - a code review should probably be done after I clean up the code.

e-koch commented on docs/dask.rst:
```diff
@@ -0,0 +1,55 @@
+
+You can optionally specify the number of threads/processes to use with ``num_workers``::
+
+    >>> with cube.use_dask_scheduler('threads', num_threads=4):  # doctest: +IGNORE_OUTPUT
```
```suggestion
    >>> with cube.use_dask_scheduler('threads', num_workers=4):  # doctest: +IGNORE_OUTPUT
```

e-koch commented on spectral_cube/io/fits.py:
```diff
@@ -226,10 +234,7 @@ def write_fits_cube(cube, filename, overwrite=False,
-            hdulist.writeto(filename, overwrite=overwrite)
-        except TypeError:
-            hdulist.writeto(filename, clobber=overwrite)
+        hdulist.writeto(filename)
```
```suggestion
        hdulist.writeto(filename, overwrite=overwrite)
```

e-koch replied:
@keflavich -- `include_origin_notes` has always been unused here. Is this still a to-do or could it be removed?

keflavich replied:
oh.  That should make this line: https://github.com/radio-astro-tools/spectral-cube/pull/618/files#diff-fa807408d14cd8bb4abf0ef50eba4a31R234 optional

astrofrog replied:
Fixed

e-koch commented:
@astrofrog -- Apart from those minor things, everything's working well in my local tests! I also get worse performance with CASA images.

I definitely agree with eventually doing temporary dumps to a zarr file. Operations like `convolve_to` to a common beam trigger the whole cube to be computed when slicing the data.

Another minor change: `DaskSpectralCube.to` could have the `self._is_huge` check removed.

astrofrog commented:
@e-koch - thanks for doing the tests! It would be great if you could let me know what minor things you ran into in case they are easily fixable here :)

@keflavich - this is ready for review! I need to think of a way to reduce duplication for ``spectral_interpolate`` (I have some ideas) but no reason that can't be done after this is merged.

e-koch commented:
@astrofrog -- Ah, I was unclear above. The comments I left were the "minor things" I meant. Otherwise things worked great!

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+                    header = self._nowcs_header
+                    # check whether the cube has beams at all
+                    # (note that "hasattr(self, 'beam') on an object with no
+                    # _beam will result in an exception....?!?!?!?)
```
Done

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+
+class FilledArrayHandler:
+    """
+    """
```
Done

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+        return SchedulerHandler(self, original_scheduler_kwargs)
+
+    def _compute(self, array):
+        # For now, always default to serial mode, but we could then expand this
+        # to allow different modes.
```
Yes, removed

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+            :class:`~spectral_cube.lower_dimensional_structures.Projection` if ``projection`` is set
+        """
+
+        # FIXME: add support for keep_shape
+        # FIXME: add support for weights
```
weights still needs adding, have opened https://github.com/radio-astro-tools/spectral-cube/issues/625

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+            :class:`~spectral_cube.lower_dimensional_structures.Projection` if ``projection`` is set
+        """
+
+        # FIXME: add support for includemask
```
https://github.com/radio-astro-tools/spectral-cube/issues/626

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+
+        if axis is None:
+            self._warn_slow('median')
+            return np.nanmedian(self._compute(data), **kwargs)
```
``NotImplementedError: The da.nanmedian function only works along an axis or a subset of axes.  The full algorithm is difficult to do in parallel`` - have added a comment

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+
+        if axis is None:
+            self._warn_slow('percentile')
+            return np.nanpercentile(data, q, **kwargs)
```
``da.nanpercentile`` doesn't exist, and no way to do it in chunks as we do when ``axis`` is specified. Added a comment.

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+        data = self._get_filled_data(fill=fill)
+
+        if rechunk is not None:
+            data = self._get_filled_data(fill=fill).rechunk(rechunk)
```
Good catch, thanks!

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+                slices.append(slice(0))
+
+        if spatial_only:
+            slices = (slice(None), slices[1], slices[2])
```
Indeed! Fixed.

astrofrog commented on spectral_cube/dask_spectral_cube.py:
```diff
@@ -0,0 +1,1278 @@
+        """
+
+        # FIXME: this does not work correctly currently due to
+        # https://github.com/dask/dask/issues/6102
```
I added a warning:

```
        warnings.warn('In some cases, the final shape of the output from downsample_axis '
                      'is incorrect, so use the result with caution', UserWarning)
```

I thought it's still worth keeping since it kind of works but might have off-by-one errors.
